cirq-core/cirq/ops/linear_combinations_test.py

```diff
@@ -1189,6 +1190,40 @@ def test_pauli_sum_formatting():
     assert str(empty) == "0.000"
 
 
+def test_pauli_sum_matrix():
+    q = cirq.LineQubit.range(3)
+    paulisum1 = cirq.X(q[0]) * cirq.X(q[1]) + 4
+    H1 = np.array(
+        [[4.0, 0.0, 0.0, 1.0], [0.0, 4.0, 1.0, 0.0], [0.0, 1.0, 4.0, 0.0], [1.0, 0.0, 0.0, 4.0]]
+    )
+    assert np.allclose(H1, paulisum1.matrix())
+    assert np.allclose(H1, paulisum1.matrix([q[0], q[1]]))
+    # Expects same matrix when change qubits order.
+    assert np.allclose(H1, paulisum1.matrix([q[1], q[0]]))
```

Comment:
This is confusion as the same matrix would be expected only in the special case where the pauli sum is symmetric wrt the qubit ordering. 

Instead of adding a special case for this, I'd recommend to only have two cases - 

1. Change qubit ordering only : Here we should also assert that `ps.with_qubits(<new order>).matrix()` is same as `ps.matrix(<new order>)`. 
2. Add new qubits : The new matrix size is larger and hence not equal to `with_qubits`.

Reply:
Thanks Tanuj. I've updated the tests to only include those two cases. But I still don't get why ps.with_qubits(<new order>).matrix() is same as ps.matrix(<new order>). Please see the example below:

In [99]: a = cirq.X(q[0]) * cirq.Y(q[1]) * cirq.Z(q[2])
In [108]: m1 = a.matrix([q[1], q[2], q[0]])
In [109]: m2 = a.with_qubits(q[1], q[2], q[0]).matrix()
In [110]: np.allclose(m1, m2)
Out[110]: False

Reply:
That's right, my bad. Please ignore my previous comment. I've also opened an issue to highlight the confusing semantics - #4270. 

I'll approve this PR to unblock you now. Apologies for the confusion which led to the long discussion.